File: frontend/src/types/portfolio.ts
/** Matches backend PortfolioSummary schema */
export interface PortfolioSummaryAPI {
  total_value: number;
  holdings_value: number;
  fd_value: number;
  total_pnl: number;
  holdings_count: number;
  fd_count: number;
  last_sync_at: string | null;
}

/** Extended summary used by the dashboard (enriched client-side) */
export interface PortfolioSummary {
  total_value: number;
  total_invested: number;
  total_pnl: number;
  pnl_percentage: number;
  holdings_value: number;
  fd_value: number;
  holdings_count: number;
  fd_count: number;
  last_synced: string | null;
  allocation_by_tag: AllocationByTag[];
  top_holdings: TopHolding[];
}

export interface AllocationByTag {
  tag_id: number | null;
  tag_name: string;
  tag_color: string | null;
  value: number;
  percentage: number;
}

export interface TopHolding {
  id: number;
  symbol: string;
  instrument_type: string;
  current_value: number;
  pnl_percentage: number;
  tags: { name: string; color: string | null }[];
}

/** Matches backend SnapshotWithReturnResponse schema */
export interface PortfolioSnapshot {
  id: number;
  date: string;
  total_value: number;
  holdings_value: number;
  fd_value: number;
  breakdown?: string | null;
  total_return?: number | null;
  created_at?: string;
}

/** Matches backend LifetimeGainsResponse schema */
export interface LifetimeGains {
  total_invested: number;
  total_withdrawn: number;
  net_invested: number;
  current_portfolio_value: number;
  lifetime_gain: number;
  lifetime_gain_pct: number | null;
  unrealized_pnl: number;
  holdings_pnl: number;
  fd_interest: number;
  realized_pnl: number;
  withdrawal_count: number;
}

File: backend/app/schemas/portfolio.py
from datetime import date, datetime
from typing import Any

from pydantic import BaseModel


class PortfolioSummary(BaseModel):
    total_value: float
    holdings_value: float
    fd_value: float
    total_pnl: float
    holdings_count: int
    fd_count: int
    last_sync_at: datetime | None = None


class SnapshotResponse(BaseModel):
    id: int
    date: date
    total_value: float
    holdings_value: float
    fd_value: float
    breakdown: str | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class SnapshotWithReturnResponse(BaseModel):
    id: int
    date: date
    total_value: float
    holdings_value: float
    fd_value: float
    breakdown: str | None = None
    total_return: float | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class LifetimeGainsResponse(BaseModel):
    total_invested: float
    total_withdrawn: float
    net_invested: float
    current_portfolio_value: float
    lifetime_gain: float
    lifetime_gain_pct: float | None = None
    unrealized_pnl: float
    holdings_pnl: float
    fd_interest: float
    realized_pnl: float
    withdrawal_count: int


class WithdrawalTimelineEntry(BaseModel):
    date: date
    amount: float
    cumulative: float
    notes: str | None = None


class BackfillRequest(BaseModel):
    kite_data: dict[str, Any]


class BackfillResponse(BaseModel):
    snapshots_created: int
    snapshots_updated: int
    message: str
    errors: list[str] = []
